test/UnitWOPIHttpRedirect.cpp

```diff
@@ -0,0 +1,151 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <config.h>
+
+#include "WopiTestServer.hpp"
+#include <Log.hpp>
+#include <Unit.hpp>
+#include <UnitHTTP.hpp>
+#include <helpers.hpp>
+
+#include <Poco/Net/HTTPRequest.h>
+
+class UnitWopiHttpRedirect : public WopiTestServer
+{
+    enum class Phase
+    {
+        Load,
+        Polling
+    } _phase;
+
+    const std::string params = "access_token=anything";
+
+protected:
+    void assertGetFileRequest(const Poco::Net::HTTPRequest&) override
+    {
+        exitTest(TestResult::Ok);
+    }
+
+public:
+    UnitWopiHttpRedirect()
+        : _phase(Phase::Load)
+    {
+    }
+
+    virtual bool handleHttpRequest(const Poco::Net::HTTPRequest& request,
+                                   Poco::MemoryInputStream& message,
+                                   std::shared_ptr<StreamSocket>& socket) override
+    {
+        Poco::URI uriReq(request.getURI());
+        Poco::RegularExpression regInfo("/wopi/files/1");
+        std::string redirectUri = "/wopi/files/0";
+        Poco::RegularExpression regRedirected(redirectUri);
+
+        Log::StreamLogger logger = Log::info();
+        if (logger.enabled())
+        {
+            logger << "Fake wopi host request URI [" << uriReq.toString() << "]:\n";
+            for (const auto& pair : request)
+            {
+                logger << '\t' << pair.first << ": " << pair.second << " / ";
+            }
+
+            LOG_END(logger, true);
+        }
+
+        // CheckFileInfo - returns redirect response
+        if (request.getMethod() == "GET" && regInfo.match(uriReq.getPath()))
```

Comment:
I'd strongly recommend using a state machine to track the phase we are in. For example, this conditional should be hit once and it should be hit first. Once hit, the phase changes and we assert that it's not called in the wrong phase. Similarly, the `if` for the new URL, it must be hit after this one, and not more than once. See UnitWOPIFileURL for an example.